internal/k8s/controllers/config_controller_test.go

```diff
@@ -0,0 +1,335 @@
+						IPAddressPools: []string{"pool3"},
+					},
+				},
+				{
+					ObjectMeta: v1.ObjectMeta{
+						Name:      "l2adv2",
+						Namespace: testNamespace,
+					},
+				},
+			},
+			PasswordSecrets: map[string]corev1.Secret{
+				"bgpsecret": {Type: corev1.SecretTypeBasicAuth, ObjectMeta: v1.ObjectMeta{Name: "bgpsecret", Namespace: testNamespace},
+					Data: map[string][]byte{"password": []byte([]byte("nopass"))}},
+			},
+			LegacyAddressPools: []v1beta1.AddressPool{
+				{
+					ObjectMeta: v1.ObjectMeta{
+						Name:      "legacypool1",
+						Namespace: testNamespace,
+					},
+					Spec: v1beta1.AddressPoolSpec{
+						Addresses: []string{
+							"10.21.0.0/16",
+							"10.51.0.0/24",
+						},
+						Protocol:      "bgp",
+						AvoidBuggyIPs: true,
+						AutoAssign:    pointer.BoolPtr(false),
+					},
+				},
+			},
+		}
+
+		calledForceReload = false
+		mockForceReload   = func() { calledForceReload = true }
+	)
+
+	if err := v1beta1.AddToScheme(scheme); err != nil {
+		t.Fatalf("add to scheme failed: %s", err)
+	}
+
+	if err := v1beta2.AddToScheme(scheme); err != nil {
+		t.Fatalf("add to scheme failed: %s", err)
+	}
+
+	if err := corev1.AddToScheme(scheme); err != nil {
+		t.Fatalf("add to scheme failed: %s", err)
+	}
+
+	tests := []struct {
+		desc         string
+		retSyncState SyncState
+	}{
+		{
+			desc:         "handler returns SyncStateSuccess",
+			retSyncState: SyncStateSuccess,
+		},
+		{
+			desc:         "handler returns SyncStateError",
+			retSyncState: SyncStateError,
+		},
+		{
+			desc:         "handler returns SyncStateErrorNoRetry",
+			retSyncState: SyncStateErrorNoRetry,
+		},
+		{
+			desc:         "handler returns SyncStateReprocessAll",
+			retSyncState: SyncStateReprocessAll,
+		},
+	}
+	for _, test := range tests {
+		calledForceReload = false
+		expectedCfg, err := config.For(resources, config.DontValidate)
+		if err != nil {
+			t.Fatalf("test %s failed, got unexpected error: %v", test.desc, err)
+		}
+
+		initObjects := objectsFromResources(resources)
+		fakeClient := fake.NewClientBuilder().WithScheme(scheme).WithObjects(initObjects...).Build()
+
+		mockHandler := func(l log.Logger, cfg *config.Config) SyncState {
+			if diff := cmp.Diff(expectedCfg, cfg, cmpopts.IgnoreUnexported(metallbcfg.Pool{})); diff != "" {
+				t.Errorf("%s: handler called with wrong resources (-want, +got)\n%s", test.desc, diff)
+			}
+			return test.retSyncState
+		}
+		r := &ConfigReconciler{
+			Client:         fakeClient,
+			Logger:         log.NewNopLogger(),
+			Scheme:         scheme,
+			Namespace:      testNamespace,
+			ValidateConfig: config.DontValidate,
+			Handler:        mockHandler,
+			ForceReload:    mockForceReload}
+
+		req := reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Namespace: testNamespace,
+			},
+		}
+
+		_, err = r.Reconcile(context.TODO(), req)
+
+		if err != nil && test.retSyncState != SyncStateError {
+			t.Fatalf("%s: shouldn't return error: %v", test.desc, err)
```

Comment:
Why the difference between fatal and errorf?

Reply:
It's my bad, wanted it to be errorf here.
however on the err := corev1.AddToScheme(scheme) I return fatal because this error is not related to the test,
but to the environment.
does it sound right to you, or do we want to always return fatal?

Reply:
Here we may want to use error, just because it's a table test. We don't want to stop at the first failure but let the other entries of the table to run.

Reply:
on fatal the test fails and stops running the TestConfigController function.
on error it fails but continues to other test cases.
so it made me believe to add this separation between test-logic fail (errorf) and env-related fail (fatal)